Mathlib/Data/List/Intervals.lean

```diff
@@ -0,0 +1,262 @@
+`multiset.Ico` and `finset.Ico` for `n ≤ x < m` as a multiset or as a finset.
+ -/
+def Ico (n m : ℕ) : List ℕ :=
+  range' n (m - n)
+#align list.Ico List.Ico
+
+namespace Ico
+
+theorem zero_bot (n : ℕ) : Ico 0 n = range n := by rw [Ico, tsub_zero, range_eq_range']
+#align list.Ico.zero_bot List.Ico.zero_bot
+
+@[simp]
+theorem length (n m : ℕ) : length (Ico n m) = m - n :=
+  by
+  dsimp [Ico]
+  simp only [length_range']
+#align list.Ico.length List.Ico.length
+
+theorem pairwise_lt (n m : ℕ) : Pairwise (· < ·) (Ico n m) :=
+  by
+  dsimp [Ico]
+  simp only [pairwise_lt_range']
+#align list.Ico.pairwise_lt List.Ico.pairwise_lt
+
+theorem nodup (n m : ℕ) : Nodup (Ico n m) :=
+  by
+  dsimp [Ico]
+  simp only [nodup_range']
+#align list.Ico.nodup List.Ico.nodup
+
+@[simp]
+theorem mem {n m l : ℕ} : l ∈ Ico n m ↔ n ≤ l ∧ l < m :=
+  by
+  suffices n ≤ l ∧ l < n + (m - n) ↔ n ≤ l ∧ l < m by simp [Ico, this]
+  cases' le_total n m with hnm hmn
+  · rw [add_tsub_cancel_of_le hnm]
+  · rw [tsub_eq_zero_iff_le.mpr hmn, add_zero]
+    exact
+      and_congr_right fun hnl =>
+        Iff.intro (fun hln => (not_le_of_gt hln hnl).elim) fun hlm => lt_of_lt_of_le hlm hmn
+#align list.Ico.mem List.Ico.mem
+
+theorem eq_nil_of_le {n m : ℕ} (h : m ≤ n) : Ico n m = [] := by
+  simp [Ico, tsub_eq_zero_iff_le.mpr h]
+#align list.Ico.eq_nil_of_le List.Ico.eq_nil_of_le
+
+theorem map_add (n m k : ℕ) : (Ico n m).map ((· + ·) k) = Ico (n + k) (m + k) := by
+  rw [Ico, Ico, map_add_range', add_tsub_add_eq_tsub_right, add_comm n k]
+#align list.Ico.map_add List.Ico.map_add
+
+theorem map_sub (n m k : ℕ) (h₁ : k ≤ n) : ((Ico n m).map fun x => x - k) = Ico (n - k) (m - k) :=
+  by rw [Ico, Ico, tsub_tsub_tsub_cancel_right h₁, map_sub_range' _ _ _ h₁]
+#align list.Ico.map_sub List.Ico.map_sub
+
+@[simp]
+theorem self_empty {n : ℕ} : Ico n n = [] :=
+  eq_nil_of_le (le_refl n)
+#align list.Ico.self_empty List.Ico.self_empty
+
+@[simp]
+theorem eq_empty_iff {n m : ℕ} : Ico n m = [] ↔ m ≤ n :=
+  Iff.intro (fun h => tsub_eq_zero_iff_le.mp <| by rw [← length, h, List.length]) eq_nil_of_le
+#align list.Ico.eq_empty_iff List.Ico.eq_empty_iff
+
+theorem append_consecutive {n m l : ℕ} (hnm : n ≤ m) (hml : m ≤ l) : Ico n m ++ Ico m l = Ico n l :=
+  by
+  dsimp only [Ico]
+  convert range'_append n (m-n) (l-m)
+  · exact (add_tsub_cancel_of_le hnm).symm
+  · rwa [← add_tsub_assoc_of_le hnm, tsub_add_cancel_of_le]
+#align list.Ico.append_consecutive List.Ico.append_consecutive
+
+@[simp]
+theorem inter_consecutive (n m l : ℕ) : Ico n m ∩ Ico m l = [] :=
+  by
+  apply eq_nil_iff_forall_not_mem.2
+  intro a
+  simp only [and_imp, not_and, not_lt, List.mem_inter, List.Ico.mem]
+  intro _ h₂ h₃
+  exfalso
+  exact not_lt_of_ge h₃ h₂
+#align list.Ico.inter_consecutive List.Ico.inter_consecutive
+
+@[simp]
+theorem bagInter_consecutive (n m l : Nat) : @List.bagInter ℕ instBEq (Ico n m) (Ico m l) = [] :=
```

Comment:
can someone just double check this theorem still says what the original mathlib intended it to say? I needed to help it along with the implicits.